diff -ruaN a/drivers/kernelsu/feature/sucompat.c b/drivers/kernelsu/feature/sucompat.c
--- a/drivers/kernelsu/feature/sucompat.c	2026-04-05 00:59:21.956528208 +0800
+++ b/drivers/kernelsu/feature/sucompat.c	2026-04-05 13:22:35.087018833 +0800
@@ -340,10 +340,11 @@
 #endif
 
 // dead code
-int __maybe_unused ksu_handle_devpts(struct inode *inode)
-{
-    return 0;
-}
+//int __maybe_unused ksu_handle_devpts(struct inode *inode)
+//{
+//    return 0;
+//}
+// Duplicate removed
 
 // sucompat: permitted process can execute 'su' to gain root access.
 void __init ksu_sucompat_init(void)
diff -ruaN a/drivers/kernelsu/include/uapi/supercall.h b/drivers/kernelsu/include/uapi/supercall.h
--- a/drivers/kernelsu/include/uapi/supercall.h	2026-04-05 00:56:42.568528269 +0800
+++ b/drivers/kernelsu/include/uapi/supercall.h	2026-04-05 13:13:03.079019051 +0800
@@ -14,7 +14,7 @@
 
 /* Magic numbers for reboot hook to install fd */
 DECLARE(__u32, KSU_INSTALL_MAGIC1, 0xDEADBEEF);
-DECLARE(__u32, KSU_INSTALL_MAGIC1, 0xDEADBEEF);
+DECLARE(__u32, KSU_INSTALL_MAGIC2, 0xDEADBEEF);
 DECLARE(__u32, KSU_FULL_VERSION_STRING, 255);
 
 struct ksu_become_daemon_cmd {
diff -ruaN a/drivers/kernelsu/other/manual_su.h b/drivers/kernelsu/other/manual_su.h
--- a/drivers/kernelsu/other/manual_su.h	2026-04-05 00:57:30.512528251 +0800
+++ b/drivers/kernelsu/other/manual_su.h	2026-04-05 13:20:06.783018890 +0800
@@ -5,9 +5,10 @@
 #include <linux/sched.h>
 #include <linux/version.h>
 
-#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 7, 0)
-#define mmap_lock mmap_sem
-#endif
+//#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 7, 0)
+//#define mmap_lock mmap_sem
+//#endif
+#define mmap_lock mmap_lock
 
 #define ksu_task_is_dead(t) ((t)->exit_state != 0)